Add iot-simplified-installer and iot-qcow2 image types to Fedora

achilleas-k opened the pull request:
_Migration of https://github.com/osbuild/osbuild-composer/pull/3282._

Adds new image types to Fedora:
- iot-simplified-installer (F38+)
- iot-qcow2-image

This PR also introduces a new interface, `FilePipeline`, which is implemented by pipelines that produce a single file (typically an image file, but it could be any single file).  Pipelines that work on single files (like `XZ`, `QCOW2`, `VMDK`) can now take any `FilePipeline` as input.

teg commented on pkg/distro/fedora/package_sets.go:
```diff
@@ -177,17 +177,65 @@ func iotCommitPackageSet(t *imageType) rpmmd.PackageSet {
 		},
 	}
 
+	switch t.arch.Name() {
```
For a follow-up: this kind of logic would be perfect to have in the Platform itself.

supakeen commented on pkg/distro/fedora/distro.go:
```diff
@@ -744,6 +796,33 @@ func newDistro(version int) distro.Distro {
+			},
+			iotSimplifiedInstallerImgType,
+		)
+		aarch64.addImageTypes(
+			&platform.Aarch64{
+				BasePlatform: platform.BasePlatform{
+					FirmwarePackages: []string{
+						"uboot-images-armv8", // ??
+						"bcm283x-firmware",
+						"arm-image-installer", // ??
+					},
+				},
+				UEFIVendor: "fedora",
+			},
+			iotSimplifiedInstallerImgType,
```
Do we want to enable the firmware overlays and /boot setup for the Pi here?

achilleas-k replied:
I based these decisions entirely on the PR in composer that I moved over.  I'd ask @runcom.

supakeen commented on pkg/distro/fedora/images.go:
```diff
@@ -481,6 +516,23 @@ func iotRawImage(workload workload.Workload,
+				img.KernelOptionsAppend = append(img.KernelOptionsAppend, "ignition.config.url="+bpIgnition.FirstBoot.ProvisioningURL)
+			}
+		case platform.FORMAT_QCOW2:
+			img.IgnitionPlatform = "qemu"
```
Why doesn't this platform also get the `bpIgnition` that raw gets?

achilleas-k replied:
Similar answer: the iot-qcow2 in the original PR didn't get the `ignition.config.url` option: https://github.com/paulwhalen/osbuild-composer/blob/e1d879691eb74bde29746e3dfa20447efad70b35/internal/distro/fedora/images.go#L501
(note that in that PR the image function was completely separate and here it's shared between the two image types).

Perhaps it was an oversight.  I'll defer to @runcom again.

supakeen commented:
Just some questions happy to approve after :)

achilleas-k commented on pkg/distro/fedora/package_sets.go:
```diff
@@ -177,17 +177,65 @@ func iotCommitPackageSet(t *imageType) rpmmd.PackageSet {
+			"efibootmgr",
+			"grub2",
+			"grub2-efi-x64",
+			"iwl1000-firmware",
+			"iwl100-firmware",
+			"iwl105-firmware",
+			"iwl135-firmware",
+			"iwl2000-firmware",
+			"iwl2030-firmware",
+			"iwl3160-firmware",
+			"iwl5000-firmware",
+			"iwl5150-firmware",
+			"iwl6050-firmware",
+			"iwl7260-firmware",
```
Good catch.  Yes, let's change them.  It doesn't change the manifests.

achilleas-k commented:
> Just some questions happy to approve after :)

Confirmed with @nullr0ute that both decisions you pointed out were intentional.

thozza reviewed (dismissed):
Very nice work! I like all the refactoring and simplification in this PR.

One thing that we discussed off-channel is the difference between `Pipeline` and `TreePipeline` and that it actually don't contain all methods of the `Pipeline` interface, unlike the `FilePipeline` does...

achilleas-k commented on pkg/distro/fedora/distro.go:
```diff
@@ -656,7 +675,16 @@ func newDistro(version int) distro.Distro {
+				ImageFormat: platform.FORMAT_QCOW2,
+			},
+			BIOS:       false,
+			UEFIVendor: "fedora",
```
Very good point, yes.  Should be defined on the distro object probably like we do for the other const strings.

achilleas-k commented on pkg/distro/fedora/partition_tables.go:
```diff
@@ -273,13 +273,13 @@ var iotSimplifiedInstallerPartitionTables = distro.BasePartitionTableMap{
-				Type: disk.EFISystemPartitionGUID,
-				UUID: disk.EFISystemPartitionUUID,
+				Size:     501 * common.MebiByte, // 501 MiB
+				Type:     "06",
```
True, it's used everywhere.

achilleas-k replied:
Opened issue to add this in a follow-up.

achilleas-k commented on pkg/distro/fedora/distro.go:
```diff
@@ -157,6 +157,28 @@ var (
 
+	iotSimplifiedInstallerImgType = imageType{
+		name:        "iot-simplified-installer",
+		nameAliases: []string{"fedora-iot-simplified-installer"},
```
Yeah I was wondering about that.  Whether there's some tooling out there that is using the old aliases and somehow wants to maintain convention.  But that's not the case.

achilleas-k replied:
Removed.

achilleas-k commented on pkg/distro/fedora/package_sets.go:
```diff
@@ -177,17 +177,56 @@ func iotCommitPackageSet(t *imageType) rpmmd.PackageSet {
+			"biosdevname",
+			"efibootmgr",
+			"grub2",
+			"grub2-efi-x64",
```
Moved to platforms.

achilleas-k commented on pkg/distro/fedora/images.go:
```diff
@@ -357,12 +357,27 @@ func iotCommitImage(workload workload.Workload,
 	img.Workload = workload
 	img.OSTreeParent = parentCommit
-	img.OSVersion = t.arch.distro.osVersion
+	img.OSVersion = d.osVersion
```
Opened issue to add this in a follow-up.

thozza reviewed (approved):
Thanks for the changes!

I'm sorry, but I found some additional small things :innocent: . Nevertheless, all of them can be changed in a follow-up :wink:

achilleas-k commented on pkg/distro/fedora/distro.go:
```diff
@@ -182,6 +203,25 @@ var (
 
+	iotQcow2ImgType = imageType{
+		name:        "iot-qcow2-image",
+		nameAliases: []string{"fedora-iot-qcow2-image"},
```
I thought I looked for all the newly added ones but I guess I was sloppy.  Thanks :)

achilleas-k commented:
Thanks!  I think I caught them all now :)

achilleas-k commented:
IoT installer failed to build yesterday because of missing `biosdevname`.  Looking into that.

thozza commented on pkg/distro/fedora/distro.go:
```diff
@@ -744,6 +788,40 @@ func newDistro(version int) distro.Distro {
+					ImageFormat: platform.FORMAT_RAW,
+					FirmwarePackages: []string{
+						"biosdevname",
+						"grub2",
```
ditto

thozza commented on pkg/distro/fedora/distro.go:
```diff
@@ -601,18 +640,11 @@ func newDistro(version int) distro.Distro {
-					"iwl6000-firmware",
-					"iwl6050-firmware",
+					"biosdevname",
+					"grub2",
```
I got tripped by `grub2` being still here. After looking around, it turns out that there is no `grub2` binary RPM. It is provided by the `grub2-pc` package.

```
[root@15aa9b685f65 /]# dnf provides grub2
Last metadata expiration check: 0:00:15 ago on Wed Aug 16 09:06:47 2023.
grub2-pc-1:2.06-89.fc38.x86_64 : Bootloader with support for Linux, Multiboot, and more
Repo        : fedora
Matched from:
Provide    : grub2 = 1:2.06-89.fc38

grub2-pc-1:2.06-95.fc38.x86_64 : Bootloader with support for Linux, Multiboot, and more
Repo        : updates
Matched from:
Provide    : grub2 = 1:2.06-95.fc38
```

So, it should not need to be listed here. I also checked other distros and we don't list `grub2` in any of them (not in platforms and also not in package sets).

`biosdevname` is also an interesting story... In other distros, we list it in the image package set. I would argue that it is not a firmware, so it probably should not be here. But I'm OK with keeping it as is for now...

achilleas-k replied:
`biosdevname` should be added conditionally based on the option in the COI pipeline.

achilleas-k replied:
I did the `biosdevname` conditional and dropped `grub2`.

achilleas-k commented:
I think we can merge this now and follow-up with the things I put in the issue another time.

thozza reviewed (dismissed):
LGTM. There are still some `grub2` packages in a pipeline `FirmwarePackages` definition, but let's do that in a follow-up as part of the issue to sort out these things.

achilleas-k commented:
The rebase on `main` in the merge-queue was failing because, in this PR, I added image types to the unit tests only for F38 and F39 and then the rebase pulled in F40 and the list was wrong: https://github.com/achilleas-k/images/blob/bd7df48a51fe0ced433e24484a5689456d81bbd1/pkg/distro/fedora/distro_test.go#L559

In hindsight, the way I added those to the tests was a bit weird; I should have used the version range, but it seemed simpler at the time.

Those tests are a bit weird in general though.  For me at least, they're always busy-work, going around all the unit tests and updating static lists, that I need to do after adding an image type.  I suppose they can catch issues if you're refactoring something in a distro and accidentally drop an image type from the registry, but when adding a new one, they sort of do the opposite: if I forget to add a new image type to a distro, I'll probably forget about the unit test lists as well and the tests will pass.  The most common scenario is that I add an image type to the distro, then the unit test fails prompting me to hunt down all the lists.

achilleas-k commented:
> The rebase on `main` in the merge-queue was failing because, in this PR, I added image types to the unit tests only for F38 and F39 and then the rebase pulled in F40 and the list was wrong: https://github.com/achilleas-k/images/blob/bd7df48a51fe0ced433e24484a5689456d81bbd1/pkg/distro/fedora/distro_test.go#L559
> 
> In hindsight, the way I added those to the tests was a bit weird; I should have used the version range, but it seemed simpler at the time.
> 
> Those tests are a bit weird in general though. For me at least, they're always busy-work, going around all the unit tests and updating static lists, that I need to do after adding an image type. I suppose they can catch issues if you're refactoring something in a distro and accidentally drop an image type from the registry, but when adding a new one, they sort of do the opposite: if I forget to add a new image type to a distro, I'll probably forget about the unit test lists as well and the tests will pass. The most common scenario is that I add an image type to the distro, then the unit test fails prompting me to hunt down all the lists.

I think we can replace this with a static list outside the code and a run of `./cmd/list-images` to compare against.  It would be like the manifest checks in composer, only it's just a list of repos x arches x image types and it would take a second to run in the unit tests.
